jina/executors/encoders/audio/fairseq.py

```diff
@@ -0,0 +1,51 @@
+import numpy as np
+from .. import BaseAudioEncoder
+from ..frameworks import BaseTorchEncoder
+from ...decorators import batching, as_ndarray
+
+
+class Wav2VecSpeechEncoder(BaseTorchEncoder, BaseAudioEncoder):
+    """
+    :class:`Wav2VecSpeechEncoder` Wav2Vec-based speech encoder presented by Facebook https://github.com/pytorch/fairseq
+    that encodes an audio signal from a `Batch x Signal Length` into a `Batch * Frames x Features` ndarray.
+    """
+
+    def __init__(self, model_path: str, *args, **kwargs):
+        """
+        Wav2Vec produces a representation for each time step at a rate of 100 Hz.
+
+        :param model_path: the path of the pretrained model. The pretrained model can be downloaded at
+            https://dl.fbaipublicfiles.com/fairseq/wav2vec/wav2vec_large.pt
+        """
+        super().__init__(*args, **kwargs)
+        self.model_path = model_path
+
+    def post_init(self):
+        import torch
+        from fairseq.models.wav2vec import Wav2VecModel
+        cp = torch.load(self.model_path, map_location=torch.device('cpu'))
+        self.model = Wav2VecModel.build_model(cp['args'], task=None)
+        self.model.load_state_dict(cp['model'])
+        self.model.eval()
+
+    @batching
+    @as_ndarray
+    def encode(self, data: np.ndarray, *args, **kwargs) -> np.ndarray:
+        """
+        Segments the audio signal into frames and encodes the frames.
+
+        :param data: a `Batch x Signal Length` ndarray, where `Signal Length` is a number of samples
+        :return: a `Batch * Frames x Features` ndarray, where `Features` is a 512-dimensional vector
+        """
+        assert data.shape[1] >= 465, 'the signal must have at least 465 samples'
+        import torch
+        embeds = []
+        for signal in data:
+            signtal_tensor = torch.from_numpy(signal).unsqueeze(0)
+            with torch.no_grad():
+                features = self.model.feature_extractor(signtal_tensor)
+                embed_tensor = self.model.feature_aggregator(features)[0]
+                embed_frames = embed_tensor.numpy().T
+                embeds.extend(embed_frames.tolist())
+        result = np.array(embeds)
+        return result
```

Comment:
The output of encoding is expected to be `Batch x Features` so that each ``Chunk`` has one vector representation and thus the ``Indexer`` can index properly. However, in your case, the output is `(Batch x Frames) x Features`. I'd suggest to figure out one way to reduce the representations to the shape of `Batch x Features`, for example concatenating all the `Frames` for each ``Chunk``.

Reply:
Here are some suggestions to speed it up a little bit and make it runnable on both CPUs and GPUs.

```
class Wav2VecSpeechEncoder(BaseTorchEncoder, BaseAudioEncoder):
    """
    :class:`Wav2VecSpeechEncoder` Wav2Vec-based speech encoder presented by Facebook https://github.com/pytorch/fairseq
    that encodes an audio signal from a `Batch x Signal Length` into a `Batch * Frames x Features` ndarray.
    """

    def __init__(self, model_path: str, *args, **kwargs):
        """
        Wav2Vec produces a representation for each time step at a rate of 100 Hz.

        :param model_path: the path of the pretrained model. The pretrained model can be downloaded at
            https://dl.fbaipublicfiles.com/fairseq/wav2vec/wav2vec_large.pt
        """
        super().__init__(*args, **kwargs)
        self.model_path = model_path

    def post_init(self):
        import torch
        from fairseq.models.wav2vec import Wav2VecModel
        cp = torch.load(self.model_path, map_location=torch.device('cpu'))
        self.model = Wav2VecModel.build_model(cp['args'], task=None)
        self.model.load_state_dict(cp['model'])
        self.model.eval()
        self.to_device(self.model)
        self._sess_func = None
        self._tensor_func = torch.tensor

    @batching
    @as_ndarray
    def encode(self, data: np.ndarray, *args, **kwargs) -> np.ndarray:
        """
        Segments the audio signal into frames and encodes the frames.

        :param data: a `Batch x Signal Length` ndarray, where `Signal Length` is a number of samples
        :return: a `Batch * Frames x Features` ndarray, where `Features` is a 512-dimensional vector
        """
        assert data.shape[1] >= 465, 'the signal must have at least 465 samples'
        embeds = []
        with self.session():
            for signal in data:
                signtal_tensor = self.array2tensor(signal.reshape(1, -1))
                features = self.model.feature_extractor(signtal_tensor)
                embed_tensor = self.model.feature_aggregator(features)[0]
                embeds.append(self.tensor2array(embed_tensor))
        return np.concatenate(embeds, 1).T

    def array2tensor(self, array):
        tensor = self._tensor_func(array)
        return tensor.cuda() if self.on_gpu else tensor

    def tensor2array(self, tensor):
        return tensor.cpu().numpy() if self.on_gpu else tensor.numpy()

    @property
    def session(self):
        if self._sess_func is None:
            self._sess_func = self.get_session()
        return self._sess_func

    def get_session(self):
        from torch import no_grad
        return no_grad
```

Reply:
I've changed the `encode` method to return `Batch x Concatenated Features`.